File: bundle-deployer/src/hdx_check_dependencies.ts
// Check if required functions and dictionaries exist in Hydrolix (Production Mode)
// COMPLETE FILE - Replace entire hdx_check_dependencies.ts with this
// UPDATED: Check with underscore prefix

import type { Bundle } from "./types/bundle.ts";

const BUNDLE_TESTING_CLUSTER = Deno.env.get("BUNDLE_TESTING_CLUSTER") || "";
const ORG_UUID = "b646d78a-5fb2-4d5f-afef-b705bf185174";
const PROJ_UUID = "469dbd34-6f06-4dfe-8fd1-9adf82123ecf";
const PROJ_NAME = "sample_project";

export async function checkDependenciesExist(
  bearerToken: string,
  bundle: Bundle,
  baseDir: string
): Promise<void> {
  const missingFunctions: string[] = [];
  const missingDictionaries: string[] = [];
  const missingFiles: string[] = [];
  
  // Check functions
  if (bundle.dependencies?.hydrolix?.required_functions) {
    const functionsUrl = `https://${BUNDLE_TESTING_CLUSTER}/config/v1/orgs/${ORG_UUID}/projects/${PROJ_UUID}/functions/`;
    
    try {
      const response = await fetch(functionsUrl, {
        headers: { 'Authorization': `Bearer ${bearerToken}` },
      });
      
      if (response.ok) {
        const responseData = await response.json();
        
        let existingFunctions: Array<{ name: string }> = [];
        if (Array.isArray(responseData)) {
          existingFunctions = responseData;
        } else if (responseData?.functions && Array.isArray(responseData.functions)) {
          existingFunctions = responseData.functions;
        } else if (responseData?.data && Array.isArray(responseData.data)) {
          existingFunctions = responseData.data;
        }
        
        const existingNames = new Set(existingFunctions.map(f => f.name));
        
        for (const functionName of bundle.dependencies.hydrolix.required_functions) {
          const fullName = `${PROJ_NAME}_${functionName}`;  // sample_project_city_name
          
          if (!existingNames.has(fullName)) {
            missingFunctions.push(functionName);
          }
          
          const filePath = `${baseDir}/functions/${functionName}.json`;
          try {
            await Deno.stat(filePath);
          } catch {
            missingFiles.push(`functions/${functionName}.json`);
          }
        }
      } else {
        throw new Error(`Failed to list functions: ${response.statusText}`);
      }
    } catch (e) {
      throw new Error(`Failed to check functions: ${e instanceof Error ? e.message : String(e)}`);
    }
  }
  
  // Check dictionaries
  if (bundle.dependencies?.hydrolix?.required_dictionaries) {
    const dictsUrl = `https://${BUNDLE_TESTING_CLUSTER}/config/v1/orgs/${ORG_UUID}/projects/${PROJ_UUID}/dictionaries/`;
    
    try {
      const response = await fetch(dictsUrl, {
        headers: { 'Authorization': `Bearer ${bearerToken}` },
      });
      
      if (response.ok) {
        const responseData = await response.json();
        
        let existingDicts: Array<{ name: string }> = [];
        if (Array.isArray(responseData)) {
          existingDicts = responseData;
        } else if (responseData?.dictionaries && Array.isArray(responseData.dictionaries)) {
          existingDicts = responseData.dictionaries;
        } else if (responseData?.data && Array.isArray(responseData.data)) {
          existingDicts = responseData.data;
        }
        
        const existingNames = new Set(existingDicts.map(d => d.name));
        
        for (const dictionaryName of bundle.dependencies.hydrolix.required_dictionaries) {
          const fullName = `${PROJ_NAME}_${dictionaryName}`;  // sample_project_ua_cat_dict
          
          if (!existingNames.has(fullName)) {
            missingDictionaries.push(dictionaryName);
          }
          
          const jsonPath = `${baseDir}/dictionaries/${dictionaryName}.json`;
          try {
            await Deno.stat(jsonPath);
          } catch {
            missingFiles.push(`dictionaries/${dictionaryName}.json`);
          }
          
          const possibleExtensions = ['csv', 'yaml', 'yml', 'tsv'];
          let foundDataFile = false;
          for (const ext of possibleExtensions) {
            try {
              await Deno.stat(`${baseDir}/dictionaries/${dictionaryName}.${ext}`);
              foundDataFile = true;
              break;
            } catch {
              // Try next
            }
          }
          if (!foundDataFile) {
            missingFiles.push(`dictionaries/${dictionaryName}.[csv/yaml/yml/tsv]`);
          }
        }
      } else {
        throw new Error(`Failed to list dictionaries: ${response.statusText}`);
      }
    } catch (e) {
      throw new Error(`Failed to check dictionaries: ${e instanceof Error ? e.message : String(e)}`);
    }
  }
  
  // Report results
  const errors: string[] = [];
  
  if (missingFunctions.length > 0) {
    errors.push(`\n❌ Missing functions on cluster:`);
    missingFunctions.forEach(name => {
      errors.push(`   - ${name} (expected as: ${PROJ_NAME}_${name})`);
    });
  }
  
  if (missingDictionaries.length > 0) {
    errors.push(`\n❌ Missing dictionaries on cluster:`);
    missingDictionaries.forEach(name => {
      errors.push(`   - ${name} (expected as: ${PROJ_NAME}_${name})`);
    });
  }
  
  if (missingFiles.length > 0) {
    errors.push(`\n⚠️  Missing local definition files:`);
    missingFiles.forEach(file => {
      errors.push(`   - ${file}`);
    });
  }
  
  if (errors.length > 0) {
    errors.push('\n📋 In production mode:');
    if (missingFunctions.length > 0 || missingDictionaries.length > 0) {
      errors.push('   • Resources must exist on cluster before deployment');
      errors.push('   • Either create them manually or run without --production flag first');
    }
    if (missingFiles.length > 0) {
      errors.push('   • Local files should be included for documentation and validation');
    }
    
    throw new Error(errors.join('\n'));
  }
  
  console.log('✓ All required dependencies exist on cluster');
  console.log('✓ All required local files present');
}

File: src/hdx_check_dependencies.rs
// Check if required functions and dictionaries exist in Hydrolix (Production Mode)

use serde_json::Value;
use std::collections::HashSet;
use tokio::fs;

use crate::bundle_struct::Bundle;
use crate::BUNDLE_TESTING_CLUSTER;

const ORG_UUID: &str = "d867bf48-4281-4496-8432-a93aa989aae6";  // markeplace-dev
const ORG_UUID_SAND: &str = "b646d78a-5fb2-4d5f-afef-b705bf185174";  // partnersandbox
const PROJ_UUID: &str = "67e79a3c-f7d6-4b33-a207-fef4579a3152";  // markeplace-dev cdn_test_project
const PROJ_UUID_SAND: &str = "469dbd34-6f06-4dfe-8fd1-9adf82123ecf";  // partnersandbox
const PROJ_NAME: &str = "cdn_test_project";

pub async fn check_dependencies_exist(
    bearer_token: &str,
    bundle: &Bundle,
    base_dir: &str,
) -> Result<(), String> {
    let mut missing_functions: Vec<String> = vec![];
    let mut missing_dictionaries: Vec<String> = vec![];
    let mut missing_files: Vec<String> = vec![];

    // Check functions
    if let Some(deps) = &bundle.dependencies {
        if let Some(hydrolix) = &deps.hydrolix {
            if let Some(required_functions) = &hydrolix.required_functions {
                let functions_url = format!(
                    "https://{}/config/v1/orgs/{}/projects/{}/functions/",
                    *BUNDLE_TESTING_CLUSTER, ORG_UUID, PROJ_UUID
                );

                let client = reqwest::Client::new();
                match client
                    .get(&functions_url)
                    .header("Authorization", format!("Bearer {}", bearer_token))
                    .send()
                    .await
                {
                    Ok(response) if response.status().is_success() => {
                        let response_data: Value = response
                            .json()
                            .await
                            .map_err(|e| format!("Failed to parse functions response: {}", e))?;

                        let empty_vec = vec![];
                        let existing_functions = if response_data.is_array() {
                            response_data.as_array().unwrap()
                        } else if let Some(functions) = response_data.get("functions") {
                            functions.as_array().unwrap_or(&empty_vec)
                        } else if let Some(data) = response_data.get("data") {
                            data.as_array().unwrap_or(&empty_vec)
                        } else {
                            &empty_vec
                        };

                        let existing_names: HashSet<String> = existing_functions
                            .iter()
                            .filter_map(|f| f.get("name").and_then(|n| n.as_str()))
                            .map(|s| s.to_string())
                            .collect();

                        for function_name in required_functions {
                            let full_name = format!("{}_{}", PROJ_NAME, function_name);

                            if !existing_names.contains(&full_name) {
                                missing_functions.push(function_name.clone());
                            }

                            let file_path =
                                format!("{}/functions/{}.json", base_dir, function_name);
                            if fs::metadata(&file_path).await.is_err() {
                                missing_files.push(format!("functions/{}.json", function_name));
                            }
                        }
                    }
                    Ok(response) => {
                        return Err(format!(
                            "Failed to list functions: {}: {:?}",
                            response.status(),
                            response
                        ));
                    }
                    Err(e) => {
                        return Err(format!("Failed to check functions: {}", e));
                    }
                }
            }
        }
    }

    // Check dictionaries
    if let Some(deps) = &bundle.dependencies {
        if let Some(hydrolix) = &deps.hydrolix {
            if let Some(required_dictionaries) = &hydrolix.required_dictionaries {
                let dicts_url = format!(
                    "https://{}/config/v1/orgs/{}/projects/{}/dictionaries/",
                    *BUNDLE_TESTING_CLUSTER, ORG_UUID, PROJ_UUID
                );

                let client = reqwest::Client::new();
                match client
                    .get(&dicts_url)
                    .header("Authorization", format!("Bearer {}", bearer_token))
                    .send()
                    .await
                {
                    Ok(response) if response.status().is_success() => {
                        let response_data: Value = response
                            .json()
                            .await
                            .map_err(|e| format!("Failed to parse dictionaries response: {}", e))?;

                        let empty_vec = vec![];
                        let existing_dicts = if response_data.is_array() {
                            response_data.as_array().unwrap()
                        } else if let Some(dictionaries) = response_data.get("dictionaries") {
                            dictionaries.as_array().unwrap_or(&empty_vec)
                        } else if let Some(data) = response_data.get("data") {
                            data.as_array().unwrap_or(&empty_vec)
                        } else {
                            &empty_vec
                        };

                        let existing_names: HashSet<String> = existing_dicts
                            .iter()
                            .filter_map(|d| d.get("name").and_then(|n| n.as_str()))
                            .map(|s| s.to_string())
                            .collect();

                        for dictionary_name in required_dictionaries {
                            let full_name = format!("{}_{}", PROJ_NAME, dictionary_name);

                            if !existing_names.contains(&full_name) {
                                missing_dictionaries.push(dictionary_name.clone());
                            }

                            let json_path =
                                format!("{}/dictionaries/{}.json", base_dir, dictionary_name);
                            if fs::metadata(&json_path).await.is_err() {
                                missing_files
                                    .push(format!("dictionaries/{}.json", dictionary_name));
                            }

                            let possible_extensions = vec!["csv", "yaml", "yml", "tsv"];
                            let mut found_data_file = false;
                            for ext in possible_extensions {
                                let data_path = format!(
                                    "{}/dictionaries/{}.{}",
                                    base_dir, dictionary_name, ext
                                );
                                if fs::metadata(&data_path).await.is_ok() {
                                    found_data_file = true;
                                    break;
                                }
                            }
                            if !found_data_file {
                                missing_files.push(format!(
                                    "dictionaries/{}.[csv/yaml/yml/tsv]",
                                    dictionary_name
                                ));
                            }
                        }
                    }
                    Ok(response) => {
                        return Err(format!(
                            "Failed to list dictionaries: {}",
                            response.status()
                        ));
                    }
                    Err(e) => {
                        return Err(format!("Failed to check dictionaries: {}", e));
                    }
                }
            }
        }
    }

    // Report results
    let mut errors: Vec<String> = vec![];

    if !missing_functions.is_empty() {
        errors.push("\n❌ Missing functions on cluster:".to_string());
        for name in &missing_functions {
            errors.push(format!(
                "   - {} (expected as: {}_{})",
                name, PROJ_NAME, name
            ));
        }
    }

    if !missing_dictionaries.is_empty() {
        errors.push("\n❌ Missing dictionaries on cluster:".to_string());
        for name in &missing_dictionaries {
            errors.push(format!(
                "   - {} (expected as: {}_{})",
                name, PROJ_NAME, name
            ));
        }
    }

    if !missing_files.is_empty() {
        errors.push("\n⚠️  Missing local definition files:".to_string());
        for file in &missing_files {
            errors.push(format!("   - {}", file));
        }
    }

    if !errors.is_empty() {
        errors.push("\n📋 In production mode:".to_string());
        if !missing_functions.is_empty() || !missing_dictionaries.is_empty() {
            errors.push("   • Resources must exist on cluster before deployment".to_string());
            errors.push(
                "   • Either create them manually or run without --production flag first"
                    .to_string(),
            );
        }
        if !missing_files.is_empty() {
            errors.push(
                "   • Local files should be included for documentation and validation".to_string(),
            );
        }

        return Err(errors.join("\n"));
    }

    println!("✓ All required dependencies exist on cluster");
    println!("✓ All required local files present");
    Ok(())
}
